builder.go

```diff
@@ -253,24 +254,46 @@ func (k *kResolver) makeAddresses(e EndpointSlice) ([]resolver.Address, string)
 	return newAddrs, ""
 }
 
-func (k *kResolver) handle(e EndpointSlice) {
-	addrs, _ := k.makeAddresses(e)
-	if len(addrs) > 0 {
+func (k *kResolver) handle(e EndpointSlice, eventType EventType) {
+	switch eventType {
+	case Added:
+		k.currentState[e.Name] = e
+	case Modified:
+		k.currentState[e.Name] = e
+	case Deleted:
+		delete(k.currentState, e.Name)
+	default:
+
+	}
+	var currentState map[resolver.Address]bool
+	var keys []resolver.Address
+	for _, value := range k.currentState {
+		addresses, _ := k.makeAddresses(value)
+		for _, address := range addresses {
+			currentState[address] = true
+		}
+
+		keys = make([]resolver.Address, 0, len(currentState))
+		for key := range currentState {
+			keys = append(keys, key)
+		}
+	}
+	if len(keys) > 0 {
 		k.cc.UpdateState(resolver.State{
-			Addresses: addrs,
+			Addresses: keys,
 		})
 		k.lastUpdateUnix.Set(float64(time.Now().Unix()))
 	}
 
 	k.endpoints.Set(float64(len(e.Endpoints)))
-	k.addresses.Set(float64(len(addrs)))
+	k.addresses.Set(float64(len(keys)))
 }
 
 func (k *kResolver) resolve() {
 	list, err := getEndpointSliceList(k.k8sClient, k.target.serviceNamespace, k.target.serviceName)
 	if err == nil {
 		for _, e := range list.Items {
-			k.handle(e)
+			k.handle(e, Added)
```

Comment:
Maybe `Modified` is more suitable? IDK
```suggestion
			k.handle(e, Modified)
```